### ISSUE TITLE
postLogoutRedirectUri not working when authority url contains a tenant ID

### ISSUE THREAD

**Issue by chill0216** [AzureAD/microsoft-authentication-library-for-js#5783]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.30.0

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

1.4.9

### Public or Confidential Client?

Public

### Description

When I modify the authority value to include the tenant ID (e.g., '[https://login.microsoftonline.com/{your-tenant-id}](https://login.microsoftonline.com/%7Byour-tenant-id%7D)') and then log out, the browser fails to redirect me back to my application. However, if I use "common" in the URL, the redirect works as expected. 

I have also confirmed that the logout URL is properly set in the Azure portal, but the issue persists. 

Notably, the issue only occurs when using a tenant-specific authority URL, as it works fine with the default 'https://login.microsoftonline.com/common' URL.

### MSAL Configuration

```javascript
export const msalConfig = {
    auth: {
        clientId: "{your-client-id}",
        authority: 'https://login.microsoftonline.com/{your-tenant-id}',
        redirectUri: "/",
        postLogoutRedirectUri: "/",
    },
    cache: {
        cacheLocation: "localStorage",
        storeAuthStateInCookie: false,
    },
};
```

### Relevant Code Snippets

```javascript
const handleLogout = () =>
    instance.logoutRedirect({
        postLogoutRedirectUri: '/',
    });
```

### Reproduction Steps

1. Change the authority to 'https://login.microsoftonline.com/{your-tenant-id}' and log in using your credentials.
2. Log out and observe that you are not redirected back to the local host.
3. Update the authority to 'https://login.microsoftonline.com/common' and log in again.
4. Log out and notice that you are now redirected back to my application.

### Expected Behavior

I anticipate that when logging out with the tenant ID in the authority URL, the browser should redirect the user to the application, similar to how it behaves when the authority URL contains "common".

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome

### Source

Internal (Microsoft)

**Comment by lalimasharda** [comment]
Hey @chill0216 , 

Are you seeing any errors in the console when you use the tenant id?

**Comment by chill0216** [comment]
> Hey @chill0216 ,
> 
> Are you seeing any errors in the console when you use the tenant id?

I am not seeing any errors, no. It just logs me out stays on this screen:
<img width="1204" alt="Screen Shot 2023-03-10 at 1 11 56 PM" src="https://user-images.githubusercontent.com/7605947/224392201-17493d9f-9409-4c95-b8f9-256a7413ab37.png">

**Comment by lalimasharda** [comment]
Okay, it seems that you are stuck on the log out page and it isn't redirecting back to your application. Can you try changing your postLogoutRedirectUri to a different uri like '/logout' or '/homepage' in your application? Are you experiencing this in logout popup as well?

I do not think tenantid is causing an issue here. Especially because you are able to login with the same authority uri.

**Comment by chill0216** [comment]
@lalimasharda Yes, I attempted to modify the URI, but it was unsuccessful. I also tried this using the MSAL react sample application in this repo and the same thing happens. It doesn't redirect me back to the application unless authority is `https://login.microsoftonline.com/common`.

https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-react-samples/react-router-sample

After initiating the logoutPopup, I am logged out within the popup and directed to the "You signed out of your account" page, however, the popup remains open and does not close unless I manually close it. 

Were you able to reproduce my issue on your end?

<img width="1066" alt="Screen Shot 2023-03-10 at 2 16 52 PM" src="https://user-images.githubusercontent.com/7605947/224406926-d38bc138-d129-4a77-a823-014e46905cfa.png">

**Comment by lalimasharda** [comment]
Yes, I did try reproducing it with that sample above with the Microsoft tenant id and it worked for me. Can you please share your verbose msal logs.. maybe I can compare it with mine and see where exactly it is failing?

**Comment by chill0216** [comment]
@lalimasharda By the way, have you tested the application on Chrome? I tested it on Safari earlier and it worked without any issues. Whenever I signed out, it always redirected me back to the application. However, I noticed that when I tried it on Chrome or Firefox, the redirect feature didn't work as expected.

Yes, here are all of the logs that I see when logging out:

```
[Sat, 11 Mar 2023 19:25:35 GMT] : @azure/msal-browser@2.33.0 : Info - Emitting event: msal:logoutStart
VM30:16943 
[Sat, 11 Mar 2023 19:25:35 GMT] : @azure/msal-react@1.5.3 : Info - MsalProvider - msal:logoutStart results in setting inProgress from none to logout
VM30:16943 
[Sat, 11 Mar 2023 19:25:35 GMT] : @azure/msal-browser@2.33.0 : Info - Emitting event: msal:logoutSuccess
```

**Comment by benaguitar** [comment]
@chill0216 I have a very similar issue on the angular version. Mine works in Safari but not in chrome or Firefox. I have recently discovered if I create a new user in AD and login using that new user it redirects correctly but does not if using the main admin AD user or external users such as a gmail account that was invited. 

The other way it works is if I am in incognito mode or set third party cookies to be blocked in non incognito mode. 

I’m curious if you have the same outcome?

**Comment by chill0216** [comment]
Hey @benaguitar, I had the exact same issue as you! I just made a new user in AD, logged in, and then logged out and was redirected back to the app correctly. It just doesn't work for the primary admin user and external users. 

@lalimasharda, looks like this same bug occurs for both @benaguitar and I.

**Comment by chill0216** [comment]
Hey @sameerag , let me know if you can help us resolve this issue. Thanks

**Comment by sameerag** [comment]
@chill0216 To summarize, AAD users work,  guest users are not redirected. Correct? What do you mean by `primary admin user`?

**Comment by chill0216** [comment]
@sameerag Sorry, by primary user I mean the user account that I used to create the Azure account (screenshot 1 - cjhill0216@gmail.com). This user account does not get redirected back to the app after logging out. However, the “Delete” user (screenshot 2) that I created by selecting “New User” > “Create new user” does get redirected back to the app after logging out. 

**Screenshot 1**
![D346759D-4B62-40FC-8AB4-3A08C606DA92](https://user-images.githubusercontent.com/7605947/225319385-e870cb26-fef9-47a0-b0f8-11ebf7af1703.jpeg)

**Screenshot 2**
![69C44AB7-D691-4FAA-BEC9-61A2298C8118](https://user-images.githubusercontent.com/7605947/225319390-7e82aaee-eb1d-4122-b837-0e9d87635522.jpeg)

**Comment by sameerag** [comment]
@chill0216 Can you share the requests for both cases when you logout? I want to check the value of `postLogoutUri` in the STS request for both the cases. 

e.g.,
```
https://login.microsoftonline.com/common/oauth2/v2.0/logout?post_logout_redirect_uri=https%3A%2F%2Flocalhost%2Fmyapp%2Flogout%2F%3F&client_id=<..
```

**Comment by chill0216** [comment]
@sameerag For cjhill0216@gmail.com, it is `https://login.microsoftonline.com/37ac774b-b8b2-4d6a-bb92-92dbfb5fdf77/oauth2/v2.0/logout?post_logout_redirect_uri=http%3A%2F%2Flocalhost%3A4200%2F&client-request-id=cc0567cc-2c3e-4e80-ae6d-12cb12a5a093` (Doesn't redirect)

For delete@cjhill0216gmail.onmicrosoft.com, it is `https://login.microsoftonline.com/37ac774b-b8b2-4d6a-bb92-92dbfb5fdf77/oauth2/v2.0/logout?post_logout_redirect_uri=http%3A%2F%2Flocalhost%3A4200%2F&client-request-id=ab8c1033-f4d7-4354-8c78-bda9f8709108` (Does redirect)

**Comment by chill0216** [comment]
@sameerag Any update on this?

**Comment by sameerag** [comment]
@chill0216 This could be a service issue or you may be having a session in one case and do not in another (in which case STS deems this is not a safe route to redirect and does not redirect). To know what is what, can you please share the fiddler trace for both cases? You can send them to my email (which is in my profile).

**Comment by chill0216** [comment]
@sameerag Sent the Fiddler trace for both. Let me know if you need me to send it in a different format

**Comment by sameerag** [comment]
Thanks @chill0216 for the logs. @konstantin-msft is checking them, one of us will reach out once we have some info.

**Comment by chill0216** [comment]
Thank a lot!

**Comment by sameerag** [comment]
@hectormmg cc

**Comment by chrishillcaption** [comment]
Any update on this? @jo-arroyo

**Comment by chill0216** [comment]
Any update on this @jo-arroyo?

**Comment by tnorling** [comment]
Your logs show that the post logout redirection did not occur because there was no session found to log out of. This can happen if the tenant the user is signed into is different from the tenant you are attempting to sign out of, as would be the case for external users. Is there a reason you don't want to use the /common endpoint? That's what we would normally recommend using in all cases, including in cases where you signed in to a specific tenant.

**Comment by chill0216** [comment]
@tnorling  Hmm, what would be the case for the user account that I used to create the Azure account not redirecting back to the login page? This is not an external user. 

Also, I have tried https://login.microsoftonline.com/common as the authority, but I get an "invalid request" error message:
`errors: ["invalid_request", "invalid_request", "invalid_request"]`. I don't think I can use `/common` since this is a single-tenant app. Is that correct?

**Comment by tnorling** [comment]
Is there any more detail on the invalid_request?

**Comment by tnorling** [comment]
The fiddler trace you shared before just had the logout request and errored out due to there not being an existing session. If you can send me a fiddler trace showing both a login and the subsequent failed logout with a non-external user I can share it with our server teams and see if they know what's going on.

**Comment by chill0216** [comment]
@tnorling I sent over the logs to your email. Let me know if you need anything else.

**Comment by tnorling** [comment]
I responded over email with a few additional questions, please take a moment to answer those questions when you can. Thanks!

**Comment by chill0216** [comment]
@tnorling Ok, I will get a response to those questions over to you today.

**Comment by chill0216** [comment]
Hey, @tnorling, I'm planning to send you an email to provide a more detailed explanation of the issue I'm currently experiencing. Additionally, I'll include a screen recording to help illustrate the problem.

**Comment by lekpeng** [comment]
hi just wondering if there has been any solution/follow up to this? i think i might be facing the same problem

**Comment by gfmoreira** [comment]
If you use useCallback it will work fine.
Solved it for me, it was exactly the same problem.

```
const { instance } = useMsal();
const logout = useCallback(() => {
		instance.logoutRedirect({
			postLogoutRedirectUri: '/',
			mainWindowRedirectUri: '/',
		});
	}, [instance]);
```

**Comment by rahulhinge** [comment]
Hello,
I am too facing the same issue. Is there any solution for this issue please?

**Comment by StijnAvdL** [comment]
Hi, 
I'm also stuck on this issue. Is there any update on this? Thanks in advance!

**Comment by rahulhinge** [comment]
It is only working for me with 'common' url. Redirect url is configured on AzureBC -> Registred Application
Following is my code sample. Hope this helps.
  const { instance } = useMsal();

  const handleLogout = () =>{
    const logoutRequest = {
      authority: 'https://login.microsoftonline.com/common'
    };

   instance.logoutRedirect(logoutRequest);
  }

On logout, it asks to pick a account then it momentarily shows sreen to close all browser windows and then redirects to suggested page.
Does anyone know how to bypass 'Pick an Account' and 'Close all Browser windows' screen please?

**Comment by MirzaMerdovic** [comment]
I am experiencing the same issue as other.
I can see in console logs that logout is being executed correctly, and post logout url is correct, but redirect doesn't happen
![image](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/16725633/b62d66cd-369d-4b44-bac7-1707d7355584)

**Comment by turenc** [comment]
```
        const msalLogoutUrl: string = `${msalConfig.authority}/oauth2/v2.0/logout?post_logout_redirect_uri=${window.location.origin}${msalConfig.postLogoutRedirectUri}&client_id=${msalConfig.clientId}`;
        window.location.replace(msalLogoutUrl);
```
I am experiencing the same issue as other.
        
![image](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/15718946/288bd014-4c03-46e2-b74c-ec7989ba7999)

**Comment by Rablet** [comment]
I had the same issue with the confidential client and noticed including the login_hint query param fixed the issue for me. Might be worth trying if it's supported in the public client as well?

It's described here: https://learn.microsoft.com/en-gb/entra/identity-platform/v2-protocols-oidc#send-a-sign-out-request:~:text=logout_hint,of%20the%20logout_hint%20parameter.

**Comment by efedericomedina** [comment]
I am having this issue a well when login out with the admin account that is .gmail, but it works when I logout with a test user.

**Comment by ManelBH** [comment]
I'm also having this issue. Any news on this?

**Comment by snathanail** [comment]
So what I did to fix this was the following. I had set up my application on Microsoft Entra ID - App Registrations. In the Authentication tab, in Single-page Application section I added the logout callback URI as follows:

![image](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/59808761/a3250f40-e5b9-4a16-9e61-623fb8e457eb)

A couple of minutes later, the logout started redirecting me back to my app!

Apparently we need to "whitelist" the logout redirect url.

HTH

**Comment by KartikSoftradix** [comment]
any exact solution for this?

**Comment by oldium** [comment]
I just discovered that the redirect URI on `common` tenant is correctly read URL-encoded, while on `consumers` tenant it works only plain text, so the following works for me (you have to have the URI registered in Redirect URIs in order to see the redirect):

```
https://login.microsoftonline.com/common/oauth2/v2.0/logout?post_logout_redirect_uri=http%3A%2F%2Flocalhost%3A3000%2F
https://login.microsoftonline.com/consumers/oauth2/v2.0/logout?post_logout_redirect_uri=http://localhost:3000/
```

I was not able to get it working to logout only my app (invalidate its tokens) by adding `client_id` (as per [OpenID Connect spec](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout)) and `logout_hint` (taken from login_hint claim as per [Entra OIDC spec](https://learn.microsoft.com/en-us/entra/identity-platform/v2-protocols-oidc#send-a-sign-out-request)) query parameters to `common` URI, though. But redirect at least worked...

Note: The redirect works only when not adding further query arguments, like `logout_hint`. With `logout_hint` the account selection is skipped, but redirection does not work...

Hope it helps.

**Comment by back2Lobby** [comment]
After struggling for hours, one simple thing worked for me. Just add the post logout uri in the redirect uri section.


<img width="809" height="441" alt="Image" src="https://github.com/user-attachments/assets/6f939c9e-92a2-4cfd-a090-1506758fb05c" />

**Comment by ohel** [comment]
Same problem here suddenly. App's been running in production for two years already and nothing was changed, suddenly the redirect stopped working. Tenant ID is in URL, and the redirect URI is configured in Azure.

What's interesting, for some users who haven't refreshed their browser's cache yet, the redirect still works. If they use incognito mode, it doesn't work. Neither the app or Azure config has been modified in any way.

Edit: Had a call with one of the users. Console log and everything was identical between a working and a non-working case. The only difference is in the cookies under login.microsoftonline.com. With the redirect working, the user had the following cookies with an expiry date (not session) that are missing in incognito mode: CCState, MC1, MS0, MSFPC

**Comment by Tyno1** [comment]
Hey guys, i was blocked on this issue as well. my solution was to let Microsfot entra id handle everything in the provider configuration in your app. 

authorization:{
params: {
scope: ENTRA_ID_SCOPE
prompt: "login".  <<<<<< this will force  re-authentication on every login and ignore any SSO session.
}
}

Then use a middleware in your application that will always check if user is authenticated and route user to login page.

**Comment by ohel** [comment]
> Hey guys, i was blocked on this issue as well. my solution was to let Microsfot entra id handle everything in the provider configuration in your app.
> 
> authorization:{ params: { scope: ENTRA_ID_SCOPE prompt: "login". <<<<<< this will force re-authentication on every login and ignore any SSO session. } }
> 
> Then use a middleware in your application that will always check if user is authenticated and route user to login page.

In my case, prompt: "login" already was set up.

This whole issue feels extremely weird. Some people report adding something unrelated such as the logout_hint (or the prompt: login) solves the issue. Other people say adding those breaks the redirect. What's going on Microsoft? I'm trying to escalate the issue to Microsoft directly if I only find a suitable support contract.

**Comment by ohel** [comment]
> I had the same issue with the confidential client and noticed including the login_hint query param fixed the issue for me. Might be worth trying if it's supported in the public client as well?
> 
> It's described here: https://learn.microsoft.com/en-gb/entra/identity-platform/v2-protocols-oidc#send-a-sign-out-request:~:text=logout_hint,of%20the%20logout_hint%20parameter.

We are using a public client. I finally found the value of login_hint and configured it to logout_hint during logout. It worked, and solved the redirect not working problem for me.

**Comment by tnorling** [comment]
Seems like most of you have solved this. For reference [our logout docs](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/logout.md#logging-out-of-msal) list out the requirements that must be met for the post logout redirect to work properly. If you've done all 4 things listed there and are still seeing this issue you should raise it through [Azure Support](https://azure.microsoft.com/en-us/support/create-ticket/) so that the relevant teams get looped in.
